src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -1046,18 +1051,19 @@ private static unsafe uint HandleEventRecv(State state, ref StreamEvent evt)
                         break;
 
                     default:
-                        Debug.Assert(state.ReadState is ReadState.Aborted or ReadState.ConnectionClosed, $"Unexpected {nameof(ReadState)} '{state.ReadState}' in {nameof(HandleEventRecv)}.");
+                        Debug.Assert(state.ReadState is ReadState.Aborted or ReadState.ConnectionClosed, $"Unexpected {nameof(ReadState)} '{state.ReadState}' in {nameof(HandleEventReceive)}.");
 
                         // There was a race between a user aborting the read stream and the callback being ran.
                         // This will eat any received data.
                         return MsQuicStatusCodes.Success;
                 }
-            }
 
-            // We're completing a pending read.
-            if (shouldComplete)
-            {
-                state.ReceiveResettableCompletionSource.Complete(readLength);
+                // We're completing a pending read. This statement must be still inside a lock, otherwise another thread
+                // could see ReadState.None and access the still incomplete completion source
+                if (shouldComplete)
```

Comment:
I was trying to figure out why we complete all the completion sources outside of the lock and could only find this historical comment: https://github.com/dotnet/runtime/pull/427/commits/d028d308a051fafa2c8170a84807218d600cd4e6#r353958420
> Can we move the task completion to outside of the lock?

So I cannot tell if moving this inside lock will cause any deadlocks (we use `CompleteAsynchronously` so it shouldn't) or have some other negative impact.

Also, if we need to move this one inside the lock, are there any other places we should investigate as well?


Lastly, I'd love @CarnaViire to give this look as well, she's the one with the most knowledge about reads.

Reply:
From offline discussion: let's try to keep the completion outside of the lock and introduce extra ReadState to imitate what is happening on the Write path

Reply:
scratch that, additional state was needed in the end, so I added `PendingReadFinished`, And I also had to make ReadAsync into state machine.